### PR TITLE
Fix tnp.hypot gradients broken by #124888

### PR DESCRIPTION
**Pull request by edenfunf**
The overflow fix in #124888 made hypot differentiate through abs/max/min, which breaks the gradient in two spots:

- |x| == |y|: all the gradient goes to x. `hypot(x, 3 - 2x)` at x=1 gives 1.414 instead of -0.707.
- one input is 0: the second derivative drops to 0.

Worked fine in 2.20, so this is a regression, and it's already in r2.22.

Fix: treat the scale as a constant (`stop_gradient`). The math is the same, so forward outputs are bit-identical and the overflow fix stays. Gradients are now exactly x / hypot. At (0, 0) the gradient is 0 instead of NaN, and complex inputs work as before.

Added gradient tests (1st/2nd order, float32/64). They fail before this change and pass after; the existing hypot tests still pass.

### CODE REVIEW

**Comment by Phoebus-Liu** [comment]
Thanks for this fix — I verified it independently against nightly and it does resolve #127230. You may want to add `Fixes #127230` to the description so that issue closes automatically when this merges; right now the PR is not linked to it.

**Setup**: tf-nightly 2.22.0.dev20260905, CPU only, float64, `tf.experimental.numpy.hypot`, `hypot(1.0, 1.0)` (the `abs(x) == abs(y)` case named in the issue).

| quantity | before | after | analytic |
|---|---|---|---|
| `d/dx` | 1.41421356 | 0.70710678 | 0.70710678 |
| `d/dy` | 0.0 | 0.70710678 | 0.70710678 |
| `d2/dx2` | 0.0 | 0.35355339 | 0.35355339 |

Two things the current tests do not cover. Both pass with your patch, so these are suggestions for locking the behaviour in rather than bug reports:

**1. Broadcast.** The regression is also visible through broadcasting, and the fix handles it correctly. With `x = [3., 4., 5.]` and scalar `y = 0.`, on `sum(hypot(x, y))`:

- `d2/dy2` is `0.0` before the patch, `0.78333333` after — matching the analytic `sum(1/|x|) = 1/3 + 1/4 + 1/5`.
- With `x` of shape `(2,1)` (`[[3.],[4.]]`) and `y` of shape `(3)` (zeros), each element of `d2/dy2` is `0.58333333 = 1/3 + 1/4`, i.e. the broadcast reduction is exact.

A broadcast regression test would protect the `stop_gradient(scale)` path for the non-scalar case.

**2. Behaviour change at infinity.** For `hypot(inf, 1.0)`, `d/dy` is `0.0` before the patch and `NaN` after. The cause is the final `where_v2`: the un-taken `result` branch contains `inf/inf = NaN`, and the routed zero gradient multiplies into it. It is confined to the infinite element itself — I checked a batched case `x = [1., 3., inf], y = [1., 4., 1.]` and the finite entries keep their exact gradients (`0.70710678`, `0.6`). `hypot` is genuinely not differentiable at infinity, so `NaN` is defensible; I would only suggest pinning it with a test so the choice is intentional rather than incidental.

Both suggestions are cosmetic — the core fix looks correct to me.

*(Verified by running the scripts locally against a patched and an unpatched install; I used AI tooling to cross-check the measurements.)*

**Comment by edenfunf** [comment]
Added the broadcast test in 68827f83ca1, using your `(2,1) x (3)` case — `d2/dy2 == 1/3 + 1/4` per element. Fails before, passes after.

The other two I measure differently, so I've left the description alone for now.

**#127230.** This PR doesn't fix it. The reproducer from that issue, patched and unpatched:

```
value = 98.24762583187021   d2 = 1.72817825305946   (expected 27.39484491972613)
```

`tf.keras.ops.hypot` doesn't go through `tf.experimental.numpy.hypot` — Keras has its own copy in `keras/src/backend/tensorflow/ops/numpy.py` with the same `abs`/`maximum`/`minimum` shape, so it needs fixing there. Happy to link it if you have a repro this change flips.

**Infinity.** `d/dy`, float64, scalar and 1-element shapes alike:

| case | before | after |
| --- | --- | --- |
| `hypot(inf, 1.0)` | nan | nan |
| `hypot(1.0, inf)` | nan | nan |
| `hypot(-inf, 2.0)` | nan | nan |
| `hypot(inf, inf)` | 0.0 | nan |

So the NaN is pre-existing. Only `hypot(inf, inf)` changes, and its old `0.0` is the tie bug — the whole gradient went to x.

Pinned it anyway, good idea: the new test puts an infinite element next to finite ones and asserts the finite gradients stay exact (0.70710678, 0.6) while the infinite one is NaN.

**Review by dmiltr3 — APPROVED** [review]

**Comment by Phoebus-Liu** [comment]
Thanks, @edenfunf, for picking up the broadcast test -- glad the `(2,1) x (3)` case is now covered (fails before, passes after). Appreciate you driving #127230 to completion; nice work!

**Review by dmiltr3 — CHANGES_REQUESTED** [review]
Hi @edenfunf,

Thank you for your contribution and for implementing the gradient fixes for `tf.experimental.numpy.hypot`! The implementation and test cases have been reviewed and approved.

However, recent changes were merged to the upstream repository in `tensorflow/python/ops/numpy_ops/` (updating `np_math_ops.py` and `np_math_ops_test.py`) after this PR branch was created. Because the target files were modified upstream, the automated integration and presubmit testing pipeline encountered conflicts with out-of-date files when attempting to merge this change.

Could you please rebase your branch on the latest `master` (or merge `master` into your branch) and push the updated branch?

```bash
git fetch upstream
git checkout fix-tnp-hypot-gradient
git rebase upstream/master
git push --force-with-lease origin fix-tnp-hypot-gradient
```

Once the branch is synced with the latest upstream `master`, the automated synchronization and presubmit testing pipeline will re-evaluate the clean diff and proceed with merging.

Thanks again!
